File: implementations/python/packages/llmproxy/src/psp_cdl_llmproxy/durable.py
# SPDX-License-Identifier: Apache-2.0
"""Opt-in durable turns and explicit historical recovery; no automatic replay."""
from datetime import datetime, timezone
from types import SimpleNamespace
from psp_cdl_core import canonical_json
from psp_cdl_cdl import aggregate_capabilities, evaluate_batch
from psp_cdl_api_server.service import ServiceError, identifier
from psp_cdl_api_server.persistence import StoreError, integer
from psp_cdl_mcpproxy import binding_digest
from .loop import BufferedLlmLoop, LoopError, copy

DURABLE_LOOP_PROFILE = "PSP-LLM-DURABLE-0.1"


class LockdownError(LoopError):
    def __init__(self, response):
        super().__init__("PSP_POST_COMPLETION_LOCKDOWN")
        self.response = response


def call(fn, code="HOST_ERROR"):
    try: return fn()
    except Exception: raise LoopError(code) from None


def same(a, b): return canonical_json(a) == canonical_json(b)
def actor(p): return {k:p[k] for k in ("tenantId", "subjectId")}


class DurableLlmLoop(BufferedLlmLoop):
    _completion_policy = "lockdown"
    def _prepare_commit(self,p,binding,command,options): return command
    def _authorize_commit(self,p,binding,context,options):
        if call(lambda:self._host.authorize_transition(copy(p),copy(context))) is not True: raise LoopError("TRANSITION_DENIED")
    def _make_buffered(self,host,on_prompt,token,session,options): return BufferedLlmLoop(self._store,self._gate,host,self._provider)
    def _turn_command(self,command,buffered): return command
    def __init__(self, store, gate, host, provider, configuration=None):
        super().__init__(store, gate, host, provider)
        if not store.durable_turns or any(not callable(getattr(host, k, None)) for k in ("plan_turn", "authorize_transition", "audit", "authorize_recovery", "recovery_policy")):
            raise LoopError("INVALID_CONFIGURATION")
        if type(configuration) is not dict or set(configuration) != {"postCompletion"} or configuration["postCompletion"] != "lockdown":
            raise LoopError("UNSUPPORTED_POST_COMPLETION")

    def _identity(self, token, scope, expected=None):
        p = self._auth.authenticate(token)
        if scope not in p["scopes"] or expected and any(p[k] != expected[k] for k in ("tenantId", "subjectId")): raise LoopError("FORBIDDEN")
        if scope == "sessions:write" and "models:invoke" not in p["scopes"]: raise LoopError("FORBIDDEN")
        return p

    @staticmethod
    def _boundary(fn):
        try: return fn()
        except LoopError: raise
        except ServiceError as exc: raise LoopError(exc.code if exc.code in ("UNAUTHENTICATED", "FORBIDDEN") else "HOST_ERROR") from None
        except StoreError as exc:
            raise LoopError(exc.code if exc.code in ("NOT_FOUND", "EXPIRED", "STATE_BUSY", "STATE_CONFLICT", "IDEMPOTENCY_CONFLICT", "PERSISTENCE_DENIED", "AUTHORIZATION_DENIED", "INVALID_STATE", "INVALID_TRANSITION", "STORE_BUSY") else "HOST_ERROR") from None
        except Exception: raise LoopError("HOST_ERROR") from None

    @staticmethod
    def _controls(options):
        if type(options) is not dict or not integer(options.get("deadline")) or not callable(options.get("cancelled")): raise LoopError("INVALID_REQUEST")
        return {k:options[k] for k in ("deadline", "cancelled")}

    @staticmethod
    def _present(receipt, recovered):
        return {**copy(receipt["output"]), "receipt":{"profile":DURABLE_LOOP_PROFILE, "requestId":receipt["requestId"], "sessionVersion":receipt["sessionVersion"], "status":receipt["status"], "recovered":recovered}}

    def run(self, token, session_id, request, options):
        return self._boundary(lambda:self._run_turn(token, session_id, request, options))

    def _run_turn(self, token, session_id, request, options):
        value, controls = copy(request), self._controls(options)
        if type(value) is not dict or set(value) != {"message"} or type(value["message"]) is not str or not identifier(session_id) or not identifier(options.get("requestId")) or not integer(options.get("expectedVersion")) or options["expectedVersion"] < 1 or not integer(options.get("maxSteps")) or not 1 <= options["maxSteps"] <= 32:
            raise LoopError("INVALID_REQUEST")
        options = {**controls, **{k:options[k] for k in ("maxSteps", "requestId", "expectedVersion")}}
        p = self._identity(token, "sessions:write")
        owner, input_digest = actor(p), binding_digest(value)
        self._check(controls)
        def preflight():
            session = self._store.execute(owner, {"action":"getSession", "sessionId":session_id})
            completion = session.get("llmCompletion")
            if session["status"] == "completed" and type(completion) is dict and completion.get("profile") == DURABLE_LOOP_PROFILE and completion.get("policy") == "lockdown":
                locked_at = completion.get("lockedAt")
                if not integer(locked_at) or locked_at > 253402300799: raise LoopError("INVALID_STATE")
                if call(lambda:self._host.audit(copy(p), {"signal":"post_completion_override_attempt", "sessionId":session_id, "inputDigest":input_digest, "at":self._host.now()}), "AUDIT_FAILED") is not True: raise LoopError("AUDIT_FAILED")
                stamp = datetime.fromtimestamp(locked_at, timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
                raise LockdownError({"error":"session_locked", "code":"PSP_POST_COMPLETION_LOCKDOWN", "message":"This session has concluded. No further interaction is permitted.", "session_id":session_id, "locked_at":stamp, "policy":"lockdown"})
            if session["status"] != "running": raise LoopError("INACTIVE_SESSION")
            old = None
            try: old = self._store.execute(owner, {"action":"getTurn", "sessionId":session_id, "requestId":options["requestId"]})
            except StoreError as exc:
                if exc.code != "NOT_FOUND": raise
            if old is not None:
                raise LoopError("TURN_ALREADY_COMMITTED" if old["inputDigest"] == input_digest and old["sessionVersion"] == options["expectedVersion"]+1 else "IDEMPOTENCY_CONFLICT")
            if session["version"] != options["expectedVersion"]: raise LoopError("STALE_SESSION")
            return session
        session = self._coordinator.run(owner, preflight)
        authority = self._snapshot(p, session)
        expires = min(authority["expires"], session["expiresAt"])
        self._check(controls, expires)
        captured, plan_error = {}, []
        def snapshot(live, current):
            try:
                if not same(current, session): raise LoopError("STALE_SESSION")
                a = self._host.snapshot(live, current)
                if not same(a, authority): raise LoopError("STALE_AUTHORITY")
                return a
            except Exception as exc:
                plan_error.append(exc)
                raise
        def prompt(live, binding):
            captured["prompt"] = copy(self._host.prompt(live, binding), "PROMPT_REJECTED")
            return captured["prompt"]
        def authorize_final(live, binding, data):
            try:
                if call(lambda:self._host.authorize_final(copy(live), copy(binding), copy(data))) is not True: return False
                plan = copy(call(lambda:self._host.plan_turn(copy(live), copy({**binding, "requestId":options["requestId"], "postCompletion":self._completion_policy}), copy(data))), "INVALID_PLAN")
                if type(plan) is not dict or set(plan) != {"state", "retained", "complete"} or type(plan["state"]) is not dict or type(plan["retained"]) is not dict or type(plan["complete"]) is not bool: raise LoopError("INVALID_PLAN")
                captured.update(plan=plan, binding=copy(binding), data=copy(data))
                return True
            except Exception as exc:
                plan_error.append(exc)
                return False
        wrapped = SimpleNamespace(authenticate=self._host.authenticate, now=self._host.now, snapshot=snapshot, prompt=prompt,
                                  verification=self._host.verification, policy=self._host.policy, authorize_final=authorize_final)
        buffered=self._make_buffered(wrapped,lambda prompt:captured.update(prompt=copy(prompt)),token,session,options)
        try: output = buffered.run(token, session_id, value, options)
        except Exception:
            if plan_error: raise plan_error[-1]
            raise
        if set(captured) != {"prompt", "plan", "binding", "data"}: raise LoopError("INVALID_PLAN")
        plan, binding, data, signed_prompt = (captured[k] for k in ("plan", "binding", "data", "prompt"))
        def fresh(committed_version=None):
            self._check(controls, expires)
            live = self._identity(token, "sessions:write", p)
            current = self._store.execute(owner, {"action":"getSession", "sessionId":session_id})
            if (not same(current, session)) if committed_version is None else current["version"] != committed_version: raise LoopError("STALE_SESSION")
            if not same(self._snapshot(live, current), authority): raise LoopError("STALE_AUTHORITY")
            self._verify(live, binding, signed_prompt)
            self._check(controls, expires)
        guard_error = []
        def guard(context):
            try:
                if context["replay"]: raise LoopError("TURN_ALREADY_COMMITTED")
                fresh()
                self._authorize_commit(p,binding,context,options)
                fresh()
                return True
            except Exception as exc:
                guard_error.append(exc)
                return False
        command = {"action":"commitTurn", "requestId":options["requestId"], "sessionId":session_id, "expectedVersion":options["expectedVersion"],
                   **{k:session[k] for k in ("nodeId", "nodeVersion", "policyVersion")}, **plan, "postCompletion":self._completion_policy, "inputDigest":input_digest, "output":output}
        command=self._prepare_commit(p,binding,self._turn_command(command,buffered),options)
        try: receipt = self._store.execute(owner, command, guard)
        except Exception:
            if guard_error: raise guard_error[0]
            raise
        def release():
            fresh(receipt["sessionVersion"])
            self._decide(p, {**binding, "committedVersion":receipt["sessionVersion"]}, data, "release", aggregate_capabilities(authority["releaseSources"], authority["releaseComplete"]))
            fresh(receipt["sessionVersion"])
            return self._present(receipt, False)
        return self._coordinator.run(owner, release)

    def recover(self, token, session_id, request_id, options):
        return self._boundary(lambda:self._recover(token, session_id, request_id, options))

    def _recover(self, token, session_id, request_id, options):
        controls = self._controls(options)
        if not identifier(session_id) or not identifier(request_id): raise LoopError("INVALID_REQUEST")
        p = self._identity(token, "sessions:read")
        owner = actor(p)
        self._check(controls)
        def release():
            session = self._store.execute(owner, {"action":"getSession", "sessionId":session_id})
            receipt = self._store.execute(owner, {"action":"getTurn", "sessionId":session_id, "requestId":request_id})
            authority = self._snapshot(p, session)
            expires = min(authority["expires"], session["expiresAt"])
            self._check(controls, expires)
            context = {**owner, "sessionId":session_id, "sessionVersion":session["version"], "epoch":self._store.epoch, "policyVersion":authority["policyVersion"],
                       "authorityRevision":authority["revision"], "phase":"recovery", "requestId":request_id, "receiptDigest":binding_digest(receipt), "deadline":controls["deadline"]}
            def fresh():
                live = self._identity(token, "sessions:read", p)
                if not same(self._store.execute(owner, {"action":"getSession", "sessionId":session_id}), session): raise LoopError("STALE_SESSION")
                if not same(self._snapshot(live, session), authority): raise LoopError("STALE_AUTHORITY")
                self._check(controls, expires)
            if call(lambda:self._host.authorize_recovery(copy(p), copy(context), copy(receipt))) is not True: raise LoopError("RECOVERY_DENIED")
            policy = copy(call(lambda:self._host.recovery_policy(copy(p), copy(context), copy(receipt))), "INVALID_POLICY")
            if type(policy) is not dict or set(policy) != {"bindingDigest", "resources"} or policy["bindingDigest"] != binding_digest(context) or type(policy["resources"]) is not list or not policy["resources"]: raise LoopError("INVALID_POLICY")
            capabilities = aggregate_capabilities(authority["releaseSources"], authority["releaseComplete"])
            resources = []
            for resource in policy["resources"]:
                if type(resource) is not dict: raise LoopError("INVALID_POLICY")
                resources.append({**resource, "capabilities":aggregate_capabilities([{"id":"recipient", "capabilities":capabilities}, {"id":"path", "capabilities":resource.get("capabilities")}], True)})
            decision = evaluate_batch(resources)["decision"]
            if decision == "unsupported": raise LoopError("UNSUPPORTED_POLICY")
            if decision != "allow": raise LoopError("OUTPUT_DENIED")
            fresh()
            return self._present(receipt, True)
        return self._coordinator.run(owner, release)

File: implementations/typescript/packages/llmproxy/src/durable.ts
// SPDX-License-Identifier: Apache-2.0
import {canonicalJson,record,type Envelope} from "@psp-cdl/core";
import {aggregateCapabilities,evaluateBatch,type PolicyInput} from "@psp-cdl/cdl";
import {SecurityService,ServiceError,identifier,type Principal} from "@psp-cdl/api-server";
import {WorkflowStore,StoreError,bounded,integer,type AccessContext} from "@psp-cdl/api-server/persistence";
import {McpDispatchGate,bindingDigest} from "@psp-cdl/mcpproxy";
import {BufferedLlmLoop,LoopError,type LoopHost,type LoopOptions,type ProviderRegistration} from "./loop.js";

export const DURABLE_LOOP_PROFILE="PSP-LLM-DURABLE-0.1";
export interface TurnPlan {state:Record<string,unknown>;retained:Record<string,unknown>;complete:boolean}
export interface DurableOptions extends LoopOptions {requestId:string;expectedVersion:number}
export type RecoveryOptions=Pick<LoopOptions,"deadline"|"cancelled">;
export interface DurableHost extends LoopHost {
  planTurn(principal:Principal,binding:Record<string,unknown>,data:Record<string,unknown>):TurnPlan|Promise<TurnPlan>;
  authorizeTransition(principal:Principal,context:AccessContext):boolean|Promise<boolean>;
  audit(principal:Principal,event:Record<string,unknown>):boolean|Promise<boolean>;
  authorizeRecovery(principal:Principal,binding:Record<string,unknown>,receipt:Record<string,unknown>):boolean|Promise<boolean>;
  recoveryPolicy(principal:Principal,binding:Record<string,unknown>,receipt:Record<string,unknown>):
    {bindingDigest:string;resources:PolicyInput[]}|Promise<{bindingDigest:string;resources:PolicyInput[]}>;
}
export class LockdownError extends LoopError {
  constructor(public readonly response:Record<string,unknown>) {super("PSP_POST_COMPLETION_LOCKDOWN");}
}
const fail=(code:string):never=>{throw new LoopError(code);};
function copy(v:unknown,code="INVALID_REQUEST"):any {try{return bounded(v);}catch{return fail(code);}}
async function call<T>(fn:()=>T|Promise<T>,code="HOST_ERROR"):Promise<T> {try{return await fn();}catch{return fail(code);}}
const same=(a:unknown,b:unknown)=>canonicalJson(a)===canonicalJson(b);
const actor=(p:Principal)=>({tenantId:p.tenantId,subjectId:p.subjectId});

/** Durable wrapper. A committed receipt is never an implicit permission to release it again. */
export class DurableLlmLoop extends BufferedLlmLoop {
  protected get completionPolicy():string {return "lockdown";}
  protected async prepareCommit(_p:Principal,_binding:Record<string,unknown>,command:Record<string,unknown>,_options:DurableOptions):Promise<Record<string,unknown>> {return command;}
  protected async authorizeCommit(p:Principal,_binding:Record<string,unknown>,context:AccessContext,_options:DurableOptions):Promise<void> {
    if(await call(()=>this.durableHost.authorizeTransition(copy(p),copy(context)))!==true)fail("TRANSITION_DENIED");
  }
  protected makeBuffered(host:LoopHost,_onPrompt:(prompt:Envelope)=>void,_token:unknown,_session:Record<string,unknown>,_options:DurableOptions):BufferedLlmLoop {return new BufferedLlmLoop(this.store,this.gate,host,this.provider);}
  protected turnCommand(command:Record<string,unknown>,_buffered:BufferedLlmLoop):Record<string,unknown> {return command;}
  constructor(store:WorkflowStore,gate:McpDispatchGate,private readonly durableHost:DurableHost,provider:ProviderRegistration,configuration:{postCompletion:"lockdown"}) {
    super(store,gate,durableHost,provider);
    if(!store.durableTurns||[durableHost.planTurn,durableHost.authorizeTransition,durableHost.audit,durableHost.authorizeRecovery,durableHost.recoveryPolicy].some(f=>typeof f!=="function")) fail("INVALID_CONFIGURATION");
    if(!record(configuration)||Object.keys(configuration).join(",")!=="postCompletion"||configuration.postCompletion!=="lockdown") fail("UNSUPPORTED_POST_COMPLETION");
  }
  protected async identity(token:unknown,scope:string,expected?:Principal):Promise<Principal> {
    const auth=new SecurityService({authenticate:t=>this.host.authenticate(t),now:()=>this.host.now(),resolve:()=>null});
    const p=await auth.authenticate(token);
    if(!p.scopes.includes(scope)||expected&&(p.tenantId!==expected.tenantId||p.subjectId!==expected.subjectId)) fail("FORBIDDEN");
    if(scope==="sessions:write"&&!p.scopes.includes("models:invoke")) fail("FORBIDDEN");
    return p;
  }
  protected async boundary<T>(fn:()=>Promise<T>):Promise<T> {
    try{return await fn();}catch(e){
      if(e instanceof LoopError) throw e;
      if(e instanceof ServiceError&&["UNAUTHENTICATED","FORBIDDEN"].includes(e.code)) return fail(e.code);
      if(e instanceof StoreError&&["NOT_FOUND","EXPIRED","STATE_BUSY","STATE_CONFLICT","IDEMPOTENCY_CONFLICT","PERSISTENCE_DENIED","AUTHORIZATION_DENIED","INVALID_STATE","INVALID_TRANSITION","STORE_BUSY"].includes(e.code)) return fail(e.code);
      return fail("HOST_ERROR");
    }
  }
  protected controls(options:RecoveryOptions):RecoveryOptions {
    if(!options||!integer(options.deadline)||typeof options.cancelled!=="function") fail("INVALID_REQUEST");
    return {deadline:options.deadline,cancelled:options.cancelled};
  }
  protected present(receipt:Record<string,unknown>,recovered:boolean):Record<string,unknown> {
    return {...copy(receipt.output),receipt:{profile:DURABLE_LOOP_PROFILE,requestId:receipt.requestId,sessionVersion:receipt.sessionVersion,status:receipt.status,recovered}};
  }
  override async run(token:unknown,sessionId:string,request:unknown,options:DurableOptions):Promise<Record<string,unknown>> {
    return this.boundary(async()=>{
      const input=copy(request),controls=this.controls(options);
      if(!record(input)||Object.keys(input).join(",")!=="message"||typeof input.message!=="string"||!identifier(sessionId)||!identifier(options.requestId)||!integer(options.expectedVersion)||options.expectedVersion<1||!integer(options.maxSteps)||options.maxSteps<1||options.maxSteps>32) fail("INVALID_REQUEST");
      options={...controls,maxSteps:options.maxSteps,requestId:options.requestId,expectedVersion:options.expectedVersion};
      const p=await this.identity(token,"sessions:write"),owner=actor(p),inputDigest=bindingDigest(input);
      this.check(controls);
      const session=await this.coordinator.run(owner,async()=>{
        const s=await this.store.execute(owner,{action:"getSession",sessionId});
        if(s.status==="completed"&&record(s.llmCompletion)&&s.llmCompletion.profile===DURABLE_LOOP_PROFILE&&s.llmCompletion.policy==="lockdown") {
          const lockedAt=s.llmCompletion.lockedAt;
          if(!integer(lockedAt)||lockedAt>253402300799) fail("INVALID_STATE");
          if(await call(()=>this.durableHost.audit(copy(p),{signal:"post_completion_override_attempt",sessionId,inputDigest,at:this.host.now()}),"AUDIT_FAILED")!==true) fail("AUDIT_FAILED");
          throw new LockdownError({error:"session_locked",code:"PSP_POST_COMPLETION_LOCKDOWN",message:"This session has concluded. No further interaction is permitted.",session_id:sessionId,locked_at:new Date((lockedAt as number)*1000).toISOString().replace(".000Z","Z"),policy:"lockdown"});
        }
        if(s.status!=="running") fail("INACTIVE_SESSION");
        let old:Record<string,unknown>|undefined;
        try{old=await this.store.execute(owner,{action:"getTurn",sessionId,requestId:options.requestId});}catch(e){if(!(e instanceof StoreError&&e.code==="NOT_FOUND"))throw e;}
        if(old) fail(old.inputDigest===inputDigest&&old.sessionVersion===options.expectedVersion+1?"TURN_ALREADY_COMMITTED":"IDEMPOTENCY_CONFLICT");
        if(s.version!==options.expectedVersion) fail("STALE_SESSION");
        return s;
      });
      const authority=await this.snapshot(p,session),expires=Math.min(authority.expires,session.expiresAt as number);
      this.check(controls,expires);
      let prompt:Envelope|undefined,plan:TurnPlan|undefined,binding:Record<string,unknown>|undefined,data:Record<string,unknown>|undefined,planError:unknown;
      const wrapped:LoopHost={
        authenticate:t=>this.host.authenticate(t),now:()=>this.host.now(),
        snapshot:async(live,s)=>{try{if(!same(s,session))fail("STALE_SESSION");const a=await this.host.snapshot(live,s);if(!same(a,authority))fail("STALE_AUTHORITY");return a;}catch(e){planError=e;throw e;}},
        prompt:async(live,b)=>{prompt=copy(await this.host.prompt(live,b),"PROMPT_REJECTED");return prompt!;},
        verification:(live,b)=>this.host.verification(live,b),policy:(live,b,d,phase)=>this.host.policy(live,b,d,phase),
        authorizeFinal:async(live,b,d)=>{
          try {
            if(await call(()=>this.host.authorizeFinal(copy(live),copy(b),copy(d)))!==true) return false;
            const proposed=copy(await call(()=>this.durableHost.planTurn(copy(live),copy({...b,requestId:options.requestId,postCompletion:this.completionPolicy}),copy(d))),"INVALID_PLAN");
            if(!record(proposed)||Object.keys(proposed).sort().join(",")!=="complete,retained,state"||!record(proposed.state)||!record(proposed.retained)||typeof proposed.complete!=="boolean") fail("INVALID_PLAN");
            plan=proposed as unknown as TurnPlan;binding=copy(b);data=copy(d);return true;
          }catch(e){planError=e;return false;}
        }
      };
      let output:Record<string,unknown>;
      const buffered=this.makeBuffered(wrapped,value=>{prompt=copy(value);},token,session,options);
      try{output=await buffered.run(token,sessionId,input,options);}catch(e){throw planError??e;}
      if(!plan||!binding||!data||!prompt) return fail("INVALID_PLAN");
      const fresh=async(committedVersion?:unknown)=>{
        this.check(controls,expires);
        const live=await this.identity(token,"sessions:write",p),s=await this.store.execute(owner,{action:"getSession",sessionId});
        if(committedVersion===undefined?!same(s,session):s.version!==committedVersion) fail("STALE_SESSION");
        if(!same(await this.snapshot(live,s),authority)) fail("STALE_AUTHORITY");
        await this.verify(live,binding!,prompt!);this.check(controls,expires);
      };
      let guardError:unknown;
      const command=await this.prepareCommit(p,binding,this.turnCommand({action:"commitTurn",requestId:options.requestId,sessionId,expectedVersion:options.expectedVersion,nodeId:session.nodeId,nodeVersion:session.nodeVersion,policyVersion:session.policyVersion,
        state:plan.state,retained:plan.retained,complete:plan.complete,postCompletion:this.completionPolicy,inputDigest,output},buffered),options);
      const receipt=await this.store.execute(owner,command,async context=>{
        try {
          if(context.replay) fail("TURN_ALREADY_COMMITTED");
          await fresh();
          await this.authorizeCommit(p,binding!,context,options);
          await fresh();return true;
        }catch(e){guardError=e;return false;}
      }).catch(e=>{throw guardError??e;});
      return this.coordinator.run(owner,async()=>{
        await fresh(receipt.sessionVersion);
        await this.decide(p,{...binding,committedVersion:receipt.sessionVersion},data!,"release",aggregateCapabilities(authority.releaseSources,authority.releaseComplete));
        await fresh(receipt.sessionVersion);
        return this.present(receipt,false);
      });
    });
  }
  async recover(token:unknown,sessionId:string,requestId:string,options:RecoveryOptions):Promise<Record<string,unknown>> {
    return this.boundary(async()=>{
      const controls=this.controls(options);
      if(!identifier(sessionId)||!identifier(requestId)) fail("INVALID_REQUEST");
      const p=await this.identity(token,"sessions:read"),owner=actor(p);this.check(controls);
      return this.coordinator.run(owner,async()=>{
        const session=await this.store.execute(owner,{action:"getSession",sessionId});
        const receipt=await this.store.execute(owner,{action:"getTurn",sessionId,requestId});
        const authority=await this.snapshot(p,session),expires=Math.min(authority.expires,session.expiresAt as number);
        this.check(controls,expires);
        const context={...owner,sessionId,sessionVersion:session.version,epoch:this.store.epoch,policyVersion:authority.policyVersion,authorityRevision:authority.revision,
          phase:"recovery",requestId,receiptDigest:bindingDigest(receipt),deadline:controls.deadline};
        const fresh=async()=>{
          const live=await this.identity(token,"sessions:read",p);
          if(!same(await this.store.execute(owner,{action:"getSession",sessionId}),session)) fail("STALE_SESSION");
          if(!same(await this.snapshot(live,session),authority)) fail("STALE_AUTHORITY");
          this.check(controls,expires);
        };
        if(await call(()=>this.durableHost.authorizeRecovery(copy(p),copy(context),copy(receipt)))!==true) fail("RECOVERY_DENIED");
        const policy=copy(await call(()=>this.durableHost.recoveryPolicy(copy(p),copy(context),copy(receipt))),"INVALID_POLICY");
        if(!record(policy)||Object.keys(policy).sort().join(",")!=="bindingDigest,resources"||policy.bindingDigest!==bindingDigest(context)||!Array.isArray(policy.resources)||!policy.resources.length) fail("INVALID_POLICY");
        const capabilities=aggregateCapabilities(authority.releaseSources,authority.releaseComplete);
        const resources=policy.resources.map((r:PolicyInput)=>{if(!record(r))return fail("INVALID_POLICY");return {...r,capabilities:aggregateCapabilities([{id:"recipient",capabilities},{id:"path",capabilities:r.capabilities}],true)};});
        const decision=evaluateBatch(resources).decision;
        if(decision==="unsupported") fail("UNSUPPORTED_POLICY");
        if(decision!=="allow") fail("OUTPUT_DENIED");
        await fresh();return this.present(receipt,true);
      });
    });
  }
}
